Add --enable=automation-service-broker

jmontleon opened the pull request.

jmontleon commented:
/assign @derekwaynecarr

soltysh commented on pkg/oc/bootstrap/clusteradd/components/ansible-service-broker/ansible_service_broker.go:
```diff
@@ -0,0 +1,45 @@
+		InstallTemplate: bootstrap.MustAsset("install/ansibleservicebroker/deploy-automation-broker-apb.yaml"),
+	}
+
+	err := component.MakeReady(
```
Just 
```go
return component.MakeReady(...
```

soltysh commented on install/ansibleservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,58 @@
+    namespace: "${NAMESPACE}"
+
+- apiVersion: batch/v1
+  kind: Job
```
Are you sure this is Job, looking at the `.spec` this should rather be DeploymentConfig.

soltysh replied:
Job is fine, what I'm saying is that the definition you have in the yaml is not a valid job. See https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/ for an example job, what you have in here is a mixture of a job and deployment.

jmontleon replied:
Thank you. I misunderstood. I'll clean it up.

jmontleon commented:
/test extended_conformance_install

djzager commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+        containers:
+        - image: docker.io/automationbroker/automation-broker-apb:latest
+          name: automation-broker-apb
+          command: [ "entrypoint.sh", "provision", "-e", "broker_name=automation-service-broker", "-e", "create_broker_namespace=true" ]
```
you could use `args:` instead of command so you don't have to specify `entrypoint.sh`.

djzager replied:
Is there a reason why we are going with `automation-service-broker` instead of `automation-broker` (default)?

djzager replied:
If you are taking the `$NAMESPACE` as an argument you may want to add `-e broker_namespace=${NAMESPACE}`

jmontleon replied:
ACK on first.
Second, that's what I was told to set it to to be similar to template-service-broker.
Third, as mentioned above the NAMESPACE is where the apb is running not the broker.

djzager commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+- apiVersion: batch/v1
+  kind: Job
+  metadata:
+    name: automation-broker-apb
```
You may want to specifically run this job in the `${NAMESPACE}` where the `automation-broker-apb` service account will live.

jmontleon replied:
It runs in the namespace set here:
https://github.com/openshift/origin/pull/19409/files#diff-0b8cd438a5f3db747099c64b502fe091R12
https://github.com/openshift/origin/pull/19409/files#diff-0b8cd438a5f3db747099c64b502fe091R24

jmontleon commented:
/test extended_conformance_install

jmontleon commented:
/test extended_conformance_install

jmontleon commented:
/test extended_conformance_install

jmontleon commented:
@soltysh sorry for the delay. The changes have been squashed and are passing again. Thanks!

jmontleon commented:
In testing I found the deadline was too short and the job would sometimes be ended before had a chance to complete. I've upped the time a bit to compensate.

jmontleon commented:
/test extended_conformance_install

jmontleon commented:
/test extended_conformance_install

jmontleon commented:
/test gcp

jwmatthews commented:
@deads2k @bparees @smarterclayton @derekwaynecarr would one of you have a few mins to review/approve this PR to help us get deployment of automation broker integrated with 'oc'?

deads2k commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+- description: Namespace of the project that is being deploy
+  displayname: broker client cert key
+  name: NAMESPACE
+  value: "automation-broker-apb"
```
We reserve `openshift-***`.  You should use a namespace under there.

@openshift/api-review this is our first external-ish component.  How shall we reserve names?

deads2k commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+  metadata:
+    name: automation-broker-apb
+  roleRef:
+    name: cluster-admin
```
Thank you for making this very clear.  Do you have a link to where this was litigated before?

@sdodson do you install this with ansible?  Do you grant this permission?

deads2k commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+    name: automation-broker-apb
+    namespace: "${NAMESPACE}"
+
+- apiVersion: rbac.authorization.k8s.io/v1beta1
```
use v1 if you please

deads2k commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+    namespace: "${NAMESPACE}"
+
+- apiVersion: rbac.authorization.k8s.io/v1beta1
+  kind: ClusterRoleBinding
```
This should be split into `install-rbac.yaml` since bindings are reconciled.  The rest should be in `install.yaml` for consistency.

deads2k commented on install/automationservicebroker/deploy-automation-broker-apb.yaml:
```diff
@@ -0,0 +1,45 @@
+        containers:
+        - image: docker.io/automationbroker/automation-broker-apb:latest
+          name: automation-broker-apb
+          args: [ "provision", "-e", "broker_name=automation-service-broker", "-e", "create_broker_namespace=true" ]
```
"create_broker_namespace=true" looks weird.  What does that do?

deads2k commented on pkg/oc/bootstrap/clusteradd/components/automation-service-broker/automation_service_broker.go:
```diff
@@ -0,0 +1,39 @@
+	}
+	glog.V(2).Infof("instantiating automation service broker template with parameters %v", params)
+
+	component := componentinstall.Template{
```
I think we're going to require the `WaitCondition` value.

@mfojtik probably worth updating to enforce.  I can't think of a reason not to.

deads2k commented:
I'm not familiar with what it does, but it appears to fit nicely.  On the authorship side, how was it?

@jmontleon @jwmatthews To make the boundary clear, we'll keep it compiling, but if there are functional breaks over time you'll be responsible for your component.

jmontleon commented:
@deads2k Thank you for the review. I have:
- moved the namespace to openshift-automation-service-broker
- used v1 for the clusterrolebinding resource
- renamed the template to install.yaml
- split the clusterrolebinding into a separate install-rbac.yaml template
- removed "create_broker_namespace=true" as we don't actually need it (it will create the namespace for the broker if it doesn't exist)
- add a wait condition for the job completion.

I think this addresses all your comments except possibly, "Do you have a link to where this was litigated before?"  Do we need to follow up with someone to permit the clusterrolebinding to be created? The job actually runs a containerized ansible role[1] that we created install the  automation broker on an openshift cluster. The broker itself does require a clusterrolebinding [2], and it's my understanding that we need this permission in order to create this and (possibly other resources) using the container.

We also fully expected and understand that we will be on the hook to maintain this going forward.

Re: authorship I found it pretty straight forward to look at the existing options (service-catalog, template-service-broker, etc.) and work from there. I think I had it at least functioning within a day.

1. https://github.com/automationbroker/automation-broker-apb
2. https://github.com/automationbroker/automation-broker-apb/blob/master/templates/broker-auth.clusterrolebinding.yaml

deads2k commented:
> Re: authorship I found it pretty straight forward to look at the existing options (service-catalog, template-service-broker, etc.) and work from there. I think I had it at least functioning within a day.
> 
> https://github.com/automationbroker/automation-broker-apb
> https://github.com/automationbroker/automation-broker-apb/blob/master/templates/broker-auth.clusterrolebinding.yaml

A cluster up record! :)

deads2k commented on install/automationservicebroker/install-rbac.yaml:
```diff
@@ -0,0 +1,23 @@
+  metadata:
+    name: automation-broker-apb
+  roleRef:
+    name: cluster-admin
```
Please make these permissions match https://github.com/openshift/openshift-ansible/blob/master/roles/ansible_service_broker/tasks/install.yml#L30-L90 .

deads2k replied:
hmmm... perhaps I misunderstood what's happening.  This is basically an installer pod?

deads2k commented:
Alright, I cleared up my confusion with @jwmatthews .  This gets cluster-admin because the job is actually installing the components.  A few details to note that are not bugs with cluster-up

 1. Components can be installed in any order.  There is no concept of a "prereq", so you may be called in any order.  Including in parallel with or before the service catalog.  Going forward, we're expecting operators to manage the coming and going of prereqs.

 2. Components must be idempotent.  We are likely to call you multiple times.

 3. Components are likely to be asked for a "remove" at some point.  It isn't a requirement today.

This looks like a good starting point.

/lgtm

jmontleon commented:
/test cmd

jmontleon commented:
/test unit
